### ISSUE TITLE
device ptr const override ?

### ISSUE THREAD

**Issue by vineelpratap** [arrayfire/arrayfire#2799]
Hi, 
 We use Arrayfire for ML training and here is simplified version of the code that we are trying to optimize.
```
auto A = af::randu(3, 4); // (1)
auto B = af::moddims(A, 4, 3); // (2)
const float* ptr = B.device<float>(); // (3)
CALL_CONV2D(ptr, ..... ); // 
```
and here is the memory allocations at the end of of each step 
```
Step (1)
---------------------------------------------------------
|     POINTER      |    SIZE    |  AF LOCK  | USER LOCK |
---------------------------------------------------------
|    0x4207280000  |       1 KB |       Yes |        No |
---------------------------------------------------------
Step (2)
---------------------------------------------------------
|     POINTER      |    SIZE    |  AF LOCK  | USER LOCK |
---------------------------------------------------------
|    0x4207280000  |       1 KB |       Yes |        No |
---------------------------------------------------------
Step (3)
---------------------------------------------------------
|     POINTER      |    SIZE    |  AF LOCK  | USER LOCK |
---------------------------------------------------------
|    0x4207280400  |       1 KB |       Yes |       Yes |
|    0x4207280000  |       1 KB |       Yes |        No |
```
In the above code, Arrayfire `evaluates` the array `B` when (3) is called. But it is not really necessary as `moddims` doesn't change the underlying data order and  we can directly access the array `A` for the same. 

Would it be possible to support const access to device ptr of array `B` using `B.device<const float>(); ` in which case we don't allocate memory for `B`. 

I understand that my use case is very specific and I'm open to alternate solutions. FWIW, solving this will greatly improve memory usage for my use case.

**Comment by 9prady9** [comment]
I don't think what you are asking (const float*) is same as what you are trying to achieve with the code.

Given that your `B` is just a change of dimensions while keeping the number of elements same. You can very well do the following.
```c++
auto A = af::randu(3, 4); // (1)
// auto B = af::moddims(A, 4, 3); // (2) - NOT NEEDED
const float* ptr = A.device<float>(); // (3)

// Just switch the dimensions while you are using the pointer inside your custom kernel
CALL_CONV2D(ptr, ..... ); 
```

That way, there is always single memory allocation.

I understand that in this case the moddims output might not seem to require an eval, but I am not yet sure if there are caveats to that approach. I will have to investigate about it a bit further and update the progress here.

In the meanwhile, what I suggested earlier should work fine as work around.

**Comment by vineelpratap** [comment]
@9prady9 Thanks !  

> I don't think what you are asking (const float*) is same as what you are trying to achieve with the code.
My intention was - given that we are explicitly asking `const` pointer, there is no need to require an `eval()`. Otherwise, `eval()` would be required. 

> I will have to investigate about it a bit further and update the progress here.

Sure ! For reference, my goal is to have a same behavior to https://pytorch.org/docs/stable/tensors.html#torch.Tensor.view

**Comment by 9prady9** [comment]
Okay, I understand what you mean now. 

Well, I think the semantic meaning of `eval()` is not same as what you are expecting from `const pointer`. `eval()` evaluates the internal representation of that particular `af::array` into actual memory location/values (by allocation of space if required - happens with JIT, sub-arrays etc.) which doesn't violate `const pointer` semantics of C++ I think. 

I have looked into the code and give below is the explanation for why it is done the way it behaves now.

At the point where `device()` is not called, the memory is shared between A and B. Therefore, to safely fetch pointer associated with `B`, it needs to be separate from memory owned by `A` so that any function call after `device()` call can use `af::array` A without any issues while the pointer associated by B can be safely used by custom kernels written by the user. This separation is needed for safe use of pointer by user written kernels. Hope that explains the behavior you are experiencing. 

This is not a bug. If you are fetching device pointer from ArrayFire for external use, there is no way for ArrayFire to track it. The user code has to call `B.unlock()` (`array::device()` calls `lock()` internally) once it's use with the device pointer is complete so that ArrayFire's memory manager will put in the free pool again.

Therefore, if you are not fetching the device pointer and all code runs via ArrayFire functions, A and B always share single memory location.

**Comment by vineelpratap** [comment]
@9prady9  Got it. I would like to rephrase my question then :)

What does it take for me to implement function like `af::view` which performs a shallow copy of the array but calling `dims()` could return different values.

To be more specific 
```
A = af::array(H, W); 
B = af::view(A, af::dim4(W, H));
std::cout << A.dims() ; // return H x W x 1 x 1
std::cout << B.dims() ; // return W x H x 1 x 1
A = A + 1; // modifies B as well as they share same data ptr
```

**Comment by 9prady9** [comment]
@vineelpratap I have tested your code on my develop branch, and output seems fine - B isn't modified.

```c++
        af::array A = af::constant(1.0, 5, 3);
        af::array B = af::moddims(A, af::dim4(3, 5));
        af_print(A);
        af_print(B);

        A = A + 1; // modifies B as well as they share same data ptr

        af_print(A);
        af_print(B);
````
 given below is the output
```bash
A
[5 3 1 1]
   Offset: 0
   Strides: [1 5 15 15]
    1.0000     1.0000     1.0000 
    1.0000     1.0000     1.0000 
    1.0000     1.0000     1.0000 
    1.0000     1.0000     1.0000 
    1.0000     1.0000     1.0000 

B
[3 5 1 1]
   Offset: 0
   Strides: [1 3 15 15]
    1.0000     1.0000     1.0000     1.0000     1.0000 
    1.0000     1.0000     1.0000     1.0000     1.0000 
    1.0000     1.0000     1.0000     1.0000     1.0000 

A
[5 3 1 1]
   Offset: 0
   Strides: [1 5 15 15]
    2.0000     2.0000     2.0000 
    2.0000     2.0000     2.0000 
    2.0000     2.0000     2.0000 
    2.0000     2.0000     2.0000 
    2.0000     2.0000     2.0000 

B
[3 5 1 1]
   Offset: 0
   Strides: [1 3 15 15]
    1.0000     1.0000     1.0000     1.0000     1.0000 
    1.0000     1.0000     1.0000     1.0000     1.0000 
    1.0000     1.0000     1.0000     1.0000     1.0000 

```

which version of ArrayFire are you using ?

**Comment by 9prady9** [comment]
Well, I checked cpu backend for quick testing though. Let me check CUDA backend also.

**Comment by 9prady9** [comment]
Yeah, no issues in all backends. Please let me know which version of ArrayFire are you using.

**Comment by vineelpratap** [comment]
Hi @9prady9 - Thanks for your help. I'm using AF 3.7.0 . 
Using `af::moddims` doesn't work for me because A and B won't share the same data pointer. I would like array B to be modified as well when array A is modified i.e., they should share the same data pointer always.

For example. I could do 
```
af_array A_ptr = A.get(); 
af::array B = af::array(A_ptr); 
```
Then, modifying A will automatically modify B BUT I can't have A.dims() and B.dims() return a different value. 

The behavior I would like to have -
```
A = af::constant(1.0, H, W); 
B = af::view(A, af::dim4(W, H));
std::cout << A.dims() ; // return H x W x 1 x 1
std::cout << B.dims() ; // return W x H x 1 x 1

A = A + 1;

float* A_ptr = A.device<float>();
float* B_ptr = A.device<float>();
// A_ptr and B_ptr should be the same
```
Was just wondering if such a thing is possible with Arrayfire. It looks like constructor using `af_array` comes close to having this behavior but couldn't find a way to modify dims of `af_array`. 


FWIW, here is as example from torch/numpy libraries 

```
// https://stackoverflow.com/a/54507446
>>> z = torch.zeros(3, 2)
>>> x = z.view(2, 3)
>>> z.fill_(1)
>>> x
tensor([[1., 1., 1.],
        [1., 1., 1.]])
```
It would be nice to have `af::view`, `af::reshape` functions with similar behavior.

**Comment by umar456** [comment]
Hey @vineelpratap. Have you looked into af::getRawPtr in the internal.h header. Its will return the pointer without performing an eval. See:

```
        af::array A           = af::randu(3, 4);       // (1)
        af::printMemInfo();
        af::array B           = af::moddims(A, 4, 3);  // (2)
        af::printMemInfo();
        //const float* ptr = B.device<float>();     // (3)
        const float* ptr = (float*)af::getRawPtr(B);  // (3)
        af::printMemInfo();
```

```
---------------------------------------------------------
|     POINTER      |    SIZE    |  AF LOCK  | USER LOCK |
---------------------------------------------------------
|  0x7ffa98400000  |       1 KB |       Yes |        No |
---------------------------------------------------------

---------------------------------------------------------
|     POINTER      |    SIZE    |  AF LOCK  | USER LOCK |
---------------------------------------------------------
|  0x7ffa98400000  |       1 KB |       Yes |        No |
---------------------------------------------------------

---------------------------------------------------------
|     POINTER      |    SIZE    |  AF LOCK  | USER LOCK |
---------------------------------------------------------
|  0x7ffa98400000  |       1 KB |       Yes |        No |
---------------------------------------------------------
```

**Comment by vineelpratap** [comment]
Perfect! `af::getRawPtr` works for my use. 

Thank you :)
